### PR TITLE
Add Binary entity creation mode

### PR DESCRIPTION
**Pull request by pablogs9**
## TODO
- [x] Fill client API headers
- [x] Fill client binary creation implementation
- [x] Design QoS API (see RMW)
- [x] Desing a message for requester/replier
- [x] Implement Agent creation in FastDDS middleware
- [x] Test in [general repo](https://github.com/eProsima/Micro-XRCE-DDS/pull/85)
- [ ] Test in [Agent unit test](https://github.com/eProsima/Micro-XRCE-DDS-Agent/blob/5fc587c60874a1b3ffd0959344428dc3710d1225/test/unittest/agent/AgentUnitTests.cpp#L1078)
- [x] Update [RMW to bin creation](https://github.com/micro-ROS/rmw-microxrcedds/pull/123)
- [x] Remove XML/REF option from RMW
- [x] Shared memory: Update matching engine in client

Extras:
- [x] Check QoS matching in FastDDS Middleware `matched_*_from_bin`
- [x] Implement xrce_obj to qos in FastDDS Middleware
- [x] Fill XRCE entities to QoS in agent

### CODE REVIEW

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Linux/320/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Windows/333/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Linux/321/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Windows/334/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Linux/324/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Windows/337/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Linux/326/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Windows/339/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Linux/327/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Windows/340/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Linux/329/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Windows/342/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Linux/330/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Windows/343/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Linux/345/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Windows/358/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Linux/346/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Windows/359/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Linux/350/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Windows/363/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Linux/351/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Windows/364/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Linux/354/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Windows/367/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Linux/356/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Windows/369/)

File: include/uxr/client/core/session/create_entities_bin.h [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+//==================================================================
+//                              PUBLIC
+//==================================================================
+
+/**
+ * @brief Buffers into the stream identified by `stream_id` an XRCE CREATE submessage with an XRCE Participant payload.
+ *        The submessage will be sent when `uxr_flash_output_streams` or `uxr_run_session` function are called.
+ *        As a result of the reception of this submessage, the Agent will create an XRCE Participant according to
+ *        the binary provides in the CREATE submessage.
```
Comment (Acuadros95) [inline include/uxr/client/core/session/create_entities_bin.h:34]:
```suggestion
 *        the binary provided in the CREATE submessage.
```

File: include/uxr/client/core/session/create_entities_bin.h [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+        uxrSession* session,
+        uxrStreamId stream_id,
+        uxrObjectId object_id,
+        uint16_t domain_id,
+        uint8_t mode);
+/**
+ * @brief Buffers into the stream identified by `stream_id` an XRCE CREATE submessage with an XRCE Topic payload.
+ *        The submessage will be sent when `uxr_flash_output_streams` or `uxr_run_session` function are called.
+ *        As a result of the reception of this submessage, the Agent will create an XRCE Topic according to
+ *        the binary provides in the CREATE submessage.
```
Comment (Acuadros95) [inline include/uxr/client/core/session/create_entities_bin.h:55]:
```suggestion
 *        the binary provided in the CREATE submessage.
```

File: include/uxr/client/core/session/create_entities_bin.h [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+        uxrObjectId participant_id,
+        const char* topic_name,
+        const char* type_name,
+        uint8_t mode);
+
+/**
+ * @brief Buffers into the stream identified by `stream_id` an XRCE CREATE submessage with an XRCE Subscriber payload.
+ *        The submessage will be sent when `uxr_flash_output_streams` or `uxr_run_session` function are called.
+ *        As a result of the reception of this submessage, the Agent will create an XRCE Publisher according to
+ *        the binary provides in the CREATE submessage.
```
Comment (Acuadros95) [inline include/uxr/client/core/session/create_entities_bin.h:81]:
```suggestion
 *        the binary provided in the CREATE submessage.
```

File: include/uxr/client/core/session/create_entities_bin.h [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+        uxrStreamId stream_id,
+        uxrObjectId object_id,
+        uxrObjectId participant_id,
+        uint8_t mode);
+
+/**
+ * @brief Buffers into the stream identified by `stream_id` an XRCE CREATE submessage with an XRCE Subscriber payload.
+ *        The submessage will be sent when `uxr_flash_output_streams` or `uxr_run_session` function are called.
+ *        As a result of the reception of this submessage, the Agent will create an XRCE Subscriber according to
+ *        the binary provides in the CREATE submessage.
```
Comment (Acuadros95) [inline include/uxr/client/core/session/create_entities_bin.h:103]:
```suggestion
 *        the binary provided in the CREATE submessage.
```

File: include/uxr/client/core/session/create_entities_bin.h [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+        uxrStreamId stream_id,
+        uxrObjectId object_id,
+        uxrObjectId participant_id,
+        uint8_t mode);
+
+/**
+ * @brief Buffers into the stream identified by `stream_id` an XRCE CREATE submessage with an XRCE DataWriter payload.
+ *        The submessage will be sent when `uxr_flash_output_streams` or `uxr_run_session` function are called.
+ *        As a result of the reception of this submessage, the Agent will create an XRCE DataWriter according to
+ *        the binary provides in the CREATE submessage.
```
Comment (Acuadros95) [inline include/uxr/client/core/session/create_entities_bin.h:125]:
```suggestion
 *        the binary provided in the CREATE submessage.
```

File: include/uxr/client/core/session/create_entities_bin.h [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+        bool reliable,
+        bool keep_last,
+        bool transient_local,
+        uint8_t mode);
+
+/**
+ * @brief Buffers into the stream identified by `stream_id` an XRCE CREATE submessage with an XRCE DataReader payload.
+ *        The submessage will be sent when `uxr_flash_output_streams` or `uxr_run_session` function are called.
+ *        As a result of the reception of this submessage, the Agent will create an XRCE DataReader according to
+ *        the binary provides in the CREATE submessage.
```
Comment (Acuadros95) [inline include/uxr/client/core/session/create_entities_bin.h:155]:
```suggestion
 *        the binary provided in the CREATE submessage.
```

**Review by Acuadros95 — APPROVED** [review]

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Linux/359/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Windows/372/)

**Pull request merged**